Tracking/components/PlotTrackHitResiduals.cpp

```diff
@@ -0,0 +1,97 @@
+  PlotTrackHitDistances(const std::string& name, ISvcLocator* svcLoc)
+      : Consumer(
+            name, svcLoc,
+            {
+            KeyValue("InputSimTrackerHits", "DCHCollection"),
+            KeyValue("InputTracksFromGenParticlesAssociation", "TracksFromGenParticlesAssociation"),
+            }) {}
+
+  void operator()(const edm4hep::SimTrackerHitCollection& simTrackerHits, const edm4hep::MCRecoTrackParticleAssociationCollection& trackParticleAssociations) const override {
+
+    for (const auto& trackParticleAssociation : trackParticleAssociations) {
+      auto genParticle = trackParticleAssociation.getSim();
+      auto track = trackParticleAssociation.getRec();
+      edm4hep::TrackState trackStateAtIP;
+      bool found_trackStateAtIP = false;
+      for (const auto& trackState : track.getTrackStates()) {
+        if (trackState.location == edm4hep::TrackState::AtIP) {
+          trackStateAtIP = trackState;
+          found_trackStateAtIP = true;
+        }
+      }
+      if (!found_trackStateAtIP)
+        throw std::runtime_error("No track state defined AtIP, exiting!");
+
+      // Build an helix out of the trackState
+      auto helixFromTrack = HelixClass_double();
+      helixFromTrack.Initialize_Canonical(trackStateAtIP.phi, trackStateAtIP.D0, trackStateAtIP.Z0, trackStateAtIP.omega, trackStateAtIP.tanLambda, m_Bz);
+
+      // Fill the histogram with residuals for hits attached to the same gen particle
+      for (const auto& simTrackerHit : simTrackerHits) {
+        auto simTrackerHitgenParticle = simTrackerHit.getParticle();
+        if (simTrackerHitgenParticle.getObjectID() == genParticle.getObjectID()) {
+          double simTrackerHitPosition[] = {simTrackerHit.x(), simTrackerHit.y(), simTrackerHit.z()};
+          double distances[3];
+          helixFromTrack.getDistanceToPoint(simTrackerHitPosition, distances);
+          // Distance[0] - distance in R-Phi plane, Distance[1] - distance along Z axis, Distance[2] - 3D distance
+          ++m_residualHist[distances[2]];
+        }
+      }
+    }
+    return;
+  }
+  Gaudi::Property<float> m_Bz{this, "Bz", 2., "Z component of the (assumed constant) magnetic field in Tesla."};
+  mutable Gaudi::Accumulators::Histogram<1> m_residualHist{this, "", "Track-hit Distances", {100, 0, 1, "Distance [mm];Entries"}};
+
+	StatusCode finalize() override {
+			TFile file("track_hit_distances.root", "RECREATE");
+			std::string name = "";
+			// Name that will appear in the stats table
+			std::string histName = "Distances";
+			nlohmann::json jH {m_residualHist};
+			auto [histo, dir] = Gaudi::Histograming::Sink::jsonToRootHistogram<Gaudi::Histograming::Sink::Traits<false, TH1D, 1>>(name, histName, jH[0]);
+      // Add overflow bin to the last bin
+      int lastBinIndex = histo.GetNbinsX();
+      histo.SetBinContent(lastBinIndex, histo.GetBinContent(lastBinIndex) + histo.GetBinContent(lastBinIndex + 1));
```

Comment:
Why this? You can easily show how many entries you get in the overflow and underflow bins when plotting. With this you also change the values of the mean and Std. Dev. that will be displayed and it's impossible to distinguish a histogram with many entries in the last bin from a histogram with many overflow entries.

Reply:
This is to visualize, by default the amount of hits that are far from the track. If you don't like I can remove it, not a big deal

Reply:
But this means that now the plot has to be always accompanied with information on how it was made because the last bin that is displayed means something different from all the other bins. And also using some statistical quantity that depends on the bin content will be wrong. There is a way of showing the underflow and overflow counts in the statistics box that doesn't change, for example in python:
``` python
import ROOT
import numpy as np

hist = ROOT.TH1F("hist", "Sample Histogram", 100, -1, 1)

data = np.random.normal(0, 1, 10000)
for value in data:
    hist.Fill(value)

canvas = ROOT.TCanvas("canvas", "Histogram Canvas", 800, 600)
hist.Draw()

ROOT.gStyle.SetOptStat(1111111)
hist.SetStats(1)

canvas.Update()
canvas.Draw()

input()
```
and then the last bin doesn't have to be touched and is automatically computed by ROOT histograms.

Reply:
Fair enough